apps/aecore/lib/aecore/util/state_trees.ex

```diff
@@ -0,0 +1,97 @@
+defmodule Aecore.Util.StateTrees do
+  @moduledoc """
+  Abstract module defining functions for known Tree types
+  """
+  alias MerklePatriciaTree.Trie
+
+  defmacro __using__([tree_type, stored_type]) when is_atom(tree_type) do
+    quote location: :keep do
+      @behaviour Aecore.Util.StateTrees
+
+      alias Aecore.Contract.{Contract, ContractStateTree}
+      alias Aecore.Chain.Identifier
+      alias Aecore.Naming.{Name, NameCommitment}
+      alias Aecore.Util.StateTrees
+      alias Aeutil.PatriciaMerkleTree
+      alias Aeutil.Serialization
+      alias MerklePatriciaTree.Trie
+
+      @typedoc "Hash of the tree"
+      @type hash :: binary()
+
+      @spec init_empty :: Trie.t()
+      def init_empty do
+        PatriciaMerkleTree.new(unquote(tree_type))
+      end
+
+      @spec put(Trie.t(), binary(), map()) :: Trie.t() | {:error, String.t()}
+      def put(tree, key, value) do
+        if StateTrees.valid_store_type?(value, unquote(stored_type)) do
+          serialized_state = Serialization.rlp_encode(value)
+          PatriciaMerkleTree.enter(tree, key, serialized_state)
+        else
+          {:error,
+           "#{__MODULE__}: Invalid value type: #{value.__struct__} but expected value type #{
+             unquote(stored_type)
+           }"}
+        end
+      end
+
+      @spec get(Trie.t(), binary()) :: map() | :none | {:error, String.t()}
+      def get(tree, key) do
+        case PatriciaMerkleTree.lookup(tree, key) do
```

Comment:
Why not use ``` with ``` to remove nesting? The deserialization of the value can error - we should handle this.